hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3290,9 +3412,27 @@ def construct_expr(ir: IR,
                    aggregations: LinkedList = LinkedList(Aggregation)):
     if type is None:
         return Expression(ir, None, indices, aggregations)
-    if isinstance(type, tarray) and is_numeric(type.element_type):
+    elif isinstance(type, tarray) and is_numeric(type.element_type):
         return ArrayNumericExpression(ir, type, indices, aggregations)
-    if isinstance(type, tndarray) and is_numeric(type.element_type):
+    elif isinstance(type, tarray):
+        etype = type.element_type
+        if isinstance(etype, (hl.tarray, hl.tset)):
+            while isinstance(etype, (hl.tarray, hl.tset)):
+                etype = etype.element_type
+        if isinstance(etype, hl.tstruct):
+            return ArrayStructExpression(ir, type, indices, aggregations)
+        else:
+            raise NotImplementedError(type)
```

Comment:
these not implemented errors are wrong. You want these to get to line 3440 below to go through typ_to_expr.

with this in mind, the no-elif model is way nicer since you don't need to duplicate that logic!